### DISCUSSION TITLE
Different windows credentials to connect to mssql server

### DISCUSSION THREAD

**Discussion by mikkelklitlund** [sqlalchemy/sqlalchemy#8758] [Usage Questions]
I am trying to connect to a sql server with a different windows username/password combination than on my local account. I know it is valid credentials, since i am able to login through SSMS.

I've tried the following:
`connection_string = "DRIVER={SQL Server};SERVER=server;DATABASE=db;UID=username;PWD=password"
connection_url = URL.create("mssql", query={"odbc_connect": connection_string})`

But i'm getting an error, which states that the username/password combo is incorrect: `[SQL Server]Login failed for user 'username'. (18456)` - but i have verified that the combo is correct.

It is needed since im developing a small application that needs to communicate with a sql server. The application will be distributed at my workplace, so it needs to be able to connect from various windows computers.

**Comment by zzzeek** [comment]
hi -

can you work with pyodbc first to get exactly what you need, then we can show you the SQL string.  pyodbc connect guidance is at https://learn.microsoft.com/en-us/sql/connect/python/pyodbc/step-1-configure-development-environment-for-pyodbc-python-development?source=recommendations&view=sql-server-ver16 and tips on connectivity are at https://github.com/mkleehammer/pyodbc/wiki/Connecting-to-SQL-Server-from-Windows

**Reply by mikkelklitlund** [reply]
Hi again,
I've tried the following with pyodbc:
`cnxn = pyodbc.connect("DRIVER={ODBC Driver 17 for SQL Server};SERVER=server;DATABASE=db;UID=username;PWD=password")
cursor = cnxn.cursor()`

I'm getting the same error `[SQL Server]Login failed for user 'username'. (18456)`

If im also applaying the `ENCRYPT=yes` it doesn't even establish connection

**Reply by zzzeek** [reply]
do you have a DBA that works with these things?  the db administrators would need to help get that working for you.

**Reply by mikkelklitlund** [reply]
I am able to connect through smss with the credentials. The database is only available on a local network. 

I asked the db admins to create a 'service acc', so it doesn't run over my personal windows account (i guess that doesn't work either as it will be the same result anyways).

Do you think it will be better to have an sql account created rather than a windows authenticated one?

**Reply by zzzeek** [reply]
this is outside of my experience (which is why I pointed you to pyodbc).   @gordthompson any advice?

**Reply by gordthompson** [reply]
> I am able to connect through smss

In the SSMS "Connect to Server" dialog, what do you select for the "Authentication" drop-down list: "Windows Authentication", or "SQL Server Authentication"?

Once connected, what to do get when you run

```sql
select system_user as su
```

**Reply by mikkelklitlund** [reply]
To make it work through smss, i shift + right click, and select “run as different user”. 

Then im prompted for the credentials. When entered, smss opens up as usual.

For the connect dialog i use windows authentication. (I will not be at my pc for some hours, but i will respond once i’ve tried it)

The first bit of this guide explains it pretty well: https://microsoft-bitools.blogspot.com/2017/07/ssms-snack-start-ssms-as-different-user.html?m=1

**Reply by gordthompson** [reply]
Your situation is explained here

https://stackoverflow.com/a/37702329/2144390

Note, however, that using the FreeTDS ODBC driver with SQLAlchemy's mssql dialect is not recommended (or supported).

TL;DR - Either use "Run As" (or equivalent) to run your Python app as the desired domain user, or use SQL Server Authentication (instead of Windows Authentication).

**Reply by mikkelklitlund** [reply]
Perfect, will try that.

Thank you!
